### PR TITLE
Refactoring of the very base code [WORK IN PROGRESS]

### PR DESCRIPTION
**Pull request by xpl**
This is a very preliminary work, everything is subject to change (20% done). Initially I thought of re-working the precision-related methods only (hence the branch name)... Well, I'm on the way to fixing them, anyway! It just took longer than I expected :)

### CODE REVIEW

**Comment by mkutny** [comment]
BTW, once you started refactoring I'll ask here, as it doesn't deserve a separate issue.

It always bothered me that `safeValue ( { "success": 0}, 'success')` resulted in `undefined`, it was definitely not what I would expect. I didn't like it but assumed it was by design since it used in so many places.

With refactored code, it now returns `0` as I would expect. Still, don't know if you expect it to work this way )

**Comment by kroitor** [comment]
> I didn't like it but assumed it was by design since it used in so many places.

Yep, it was by design... by a previous bad design.

> With refactored code, it now returns 0 as I would expect. Still, don't know if you expect it to work this way )

Right, as we are now going to catch those 0-cases and refix them properly, but the safe* family should still behave as you would expect it.

**Comment by kornrunner** [comment]
I reckon it doesn't fall into refactoring, but perhaps we could include metainfo for create/fetch deposit address in this sprint as well?

**Comment by kroitor** [comment]
@kornrunner will do!

File: js/base/Exchange.js [hunk 1]
```diff
@@ -80,9 +88,9 @@ module.exports = class Exchange {
 
         this.iso8601          = timestamp => new Date (timestamp).toISOString ()
         this.parse8601        = x => Date.parse (((x.indexOf ('+') >= 0) || (x.slice (-1) == 'Z')) ? x : (x + 'Z'))
-        this.milliseconds     = Date.now
-        this.microseconds     = () => Math.floor (this.milliseconds () * 1000)
-        this.seconds          = () => Math.floor (this.milliseconds () / 1000)
+        this.milliseconds     = () => Math.floor (time.now ())
+        this.microseconds     = () => Math.floor (time.now () * 1000)
+        this.seconds          = () => Math.floor (time.now () / 1000)
```
Comment (kroitor) [inline js/base/Exchange.js]: @xpl, does this line need a parseInt (for integer seconds)?

**Comment by mkutny** [comment]
One more thought on the `safe*` family.

As I've got from the previous #1250  discussion it's also a common case when exchanges wrap native json types into quotes, so these checks are needed:
```
success = this.safeValue (response, 'success')
if (typeof success === 'string') {
  if (success === 'true')
    success = true;
  else if (success === 'false')
    success = false;
  else
    throw new MalformedResponse()
}
```

What if we had a `safeBoolean` instead?

**Comment by kroitor** [comment]
> What if we had a safeBoolean instead?

Will add it as well! However, with booleans there's a little ambiguity...

What should a safeBoolean yield in these cases:

- true 
- "true"
- "ok" ?
- "1"

Need to define the rules before we enter deep waters...

**Comment by mkutny** [comment]
> What should a safeBoolean yield in these cases:

As I understand it (correct me if I'm wrong) the problem this `safe*` family solves (apart from checking for the key existence) is coercion of "quoted native types" to just native types.

So if I expect `123` but json has `"123"` then `safeInteger()` will give me `123`.

Likewise, if I expect `true` but json has `"true"` then `safeBoolean()` will give me `true`.

But if I expect `"ok"/"not ok"` then I should use `safeString()`, not `safeBoolean()`, does it make sense?

**Comment by kroitor** [comment]
@mkutny yep, it does. Will implement safeBoolean this way ) Thx!

File: js/base/Exchange.js [hunk 2]
```diff
@@ -1,24 +1,37 @@
 "use strict";
 
-//-----------------------------------------------------------------------------
+/*  ------------------------------------------------------------------------ */
 
+<<<<<<< HEAD
```
Comment (TrejGun) [inline js/base/Exchange.js]: bad merge (!)

**Comment by frosty00** [comment]
costToPrecision (symbol, cost) {
        return parseFloat (cost).toFixed (this.markets[symbol].precision.price)
    }

    priceToPrecision (symbol, price) {
        return parseFloat (price).toFixed (this.markets[symbol].precision.price)
    }

    amountToPrecision (symbol, amount) {
        return this.truncate (amount, this.markets[symbol].precision.amount)
    }

These methods require precision in markets to have an amount attribute, however this raises the following error when the precision is not populated in .markets (in Python):

     File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ccxt-1.10.755-py3.5.egg/ccxt/base/exchange.py", line 992, in calculate_fee
        cost = float(self.cost_to_precision(symbol, amount * price))
     File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ccxt-1.10.755-py3.5.egg/ccxt/base/exchange.py", line 778, in cost_to_precision
        return ('{:.' + str(self.markets[symbol]['precision']['price']) + 'f}').format(float(cost))
     KeyError: 'price'

**Pull request merged**